src/main/java/org/commonjava/util/partyline/impl/infinispan/model/InfinispanJF.java

```diff
@@ -0,0 +1,615 @@
+        }
+
+        @Override
+        public int hashCode()
+        {
+            int result = jointIdx;
+            result = 31 * result + originalThreadName.hashCode();
+            result = 31 * result + (int) ( ctorTime ^ ( ctorTime >>> 32 ) );
+            return result;
+        }
+
+        /**
+         * If this stream is in the process of closing, throw {@link IOException}. While the read-bytes count in this
+         * stream equals the flushed-bytes count in the associated output stream, wait for new content. If the output stream closes while we're
+         * waiting, return -1. If the thread is interrupted while we're waiting, return -1.
+         *
+         * If the current mapped buffer has been completely read, map the next section of content from the file. Then read the next byte from the
+         * buffer, increment the read-bytes count, and return it.
+         */
+        @Override
+        public int read() throws IOException
+        {
+            synchronized ( InfinispanJF.this )
+            {
+                if ( closed )
+                {
+                    throw new IOException( "Joint: " + jointIdx + "(" + originalThreadName
+                                                           + "): Cannot read from closed stream!" );
+                }
+
+            }
+            // Is it better to write getting logic into FileBlock rather than expose the whole buffer?
+            if ( block.getBuffer().position() == block.getBuffer().limit() )
+            {
+                try
+                {
+                    FileBlock next = filesystem.getNextBlock( block );
+                    block = next;
+                }
+                catch ( final InterruptedException e)
+                {
+                    return -1;
+                }
+
+            }
+
+            // be extra careful...if the new buffer is empty, return EOF.
+            if ( block.isEOF() )
+            {
+                return -1;
+            }
+
+            final int result = block.getBuffer().get();
+            read++;
+
+            // byte is signed in java. Converting to unsigned:
+            return result & 0xff;
+        }
+
+        /**
+         * Mark this stream as closed to no further reads can proceed. Then, call {@link InfinispanJF#jointClosed(JoinInputStream, String)} to notify the parent
+         * output stream to decrement its open-reader count and notify anyone waiting in case a close is in progress.
+         */
+        @Override
+        public void close() throws IOException
+        {
+            Logger logger = LoggerFactory.getLogger( getClass() );
+            logger.trace( "Joint: {} ({}) close() called.", jointIdx, originalThreadName );
+            if ( closed )
+            {
+                logger.trace( "Joint: {} ({}) already closed.", jointIdx, originalThreadName );
+                return;
+            }
+
+            closed = true;
```

Comment:
We should check the RandomAccessFile form of this interface to see if we have an opLock here. I'm not sure, but it looks like this could be a race condition waiting to happen.

Reply:
They're the same